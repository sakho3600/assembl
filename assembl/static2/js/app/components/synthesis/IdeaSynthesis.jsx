import React from 'react';
import { Image } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';
import { Link } from 'react-router';

import StatisticsDoughnut from '../debate/common/statisticsDoughnut';
import PostsAndContributorsCount from '../common/postsAndContributorsCount';

const TitleUnderHyphen = ({ value }) => {
  return (
    <div className="title-under-hyphen-container">
      <div className="title-hyphen">&nbsp;</div>
      <h3 className="title-under-hyphen-text">
        {value}
      </h3>
    </div>
  );
};

const SynthesisBody = ({ value }) => {
  return (
    <p className="synthesis-body">
      {value}
    </p>
  );
};

const LinkToIdea = ({ href }) => {
  return (
    <Link className="idea-link" to={href}>
      {'> '}
      <Translate value="synthesis.seeConversation" />
    </Link>
  );
};

const SynthesisStats = ({ numContributors, numPosts, ideaLink }) => {
  return (
    <div className="synthesis-stats">
      <StatisticsDoughnut elements={[]} placement="after" />
      <PostsAndContributorsCount vertical numContributors={numContributors} numPosts={numPosts} />
      <LinkToIdea href={ideaLink} />
    </div>
  );
};

const ImageWithSynthesisStats = ({ imageSrc, numContributors, numPosts, ideaLink }) => {
  return (
    <div className="synthesis-image-stats-container">
      <Image className="synthesis-image" responsive src={imageSrc} />
      <SynthesisStats numContributors={numContributors} numPosts={numPosts} ideaLink={ideaLink} />
    </div>
  );
};

export default ({ title, imageSrc, body, numContributors, numPosts, ideaLink }) => {
  return (
    <div className="idea-synthesis">
      <TitleUnderHyphen value={title} />
      <ImageWithSynthesisStats imageSrc={imageSrc} numContributors={numContributors} numPosts={numPosts} ideaLink={ideaLink} />
      <SynthesisBody value={body} />
    </div>
  );
};