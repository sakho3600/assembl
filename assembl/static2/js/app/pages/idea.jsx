import React from 'react';
import { connect } from 'react-redux';
import { Translate } from 'react-redux-i18n';
import { compose, graphql } from 'react-apollo';
import { Grid } from 'react-bootstrap';

import { updateContentLocale } from '../actions/contentLocaleActions';
import Header from '../components/debate/common/header';
import IdeaQuery from '../graphql/IdeaQuery.graphql';
import IdeaWithPostsQuery from '../graphql/IdeaWithPostsQuery.graphql';
import GoUp from '../components/common/goUp';
import Loader from '../components/common/loader';
import { getConnectedUserId } from '../utils/globalFunctions';
import Announcement from './../components/debate/common/announcement';
import ColumnsView from '../components/debate/multiColumns/columnsView';
import ThreadView from '../components/debate/thread/threadView';

export const transformPosts = (edges, messageColumns, additionnalProps = {}) => {
  const postsByParent = {};

  const columns = { null: { colColor: null, colName: null } };
  messageColumns.forEach((col) => {
    columns[col.messageClassifier] = { colColor: col.color, colName: col.name };
  });

  edges.forEach((e) => {
    const p = { ...e.node, ...additionnalProps, ...columns[e.node.messageClassifier] };
    const items = postsByParent[p.parentId] || [];
    postsByParent[p.parentId] = items;
    items.push(p);
  });

  const getChildren = (id) => {
    return (postsByParent[id] || []).map((post) => {
      const newPost = post;
      // We modify the object in place, we are sure it's already a copy from
      // the forEach edges above.
      newPost.children = getChildren(post.id);
      return newPost;
    });
  };

  // postsByParent.null is the list of top posts
  return (postsByParent.null || []).map((p) => {
    const newPost = p;
    newPost.children = getChildren(p.id);
    return newPost;
  });
};

const noRowsRenderer = () => {
  return (
    <div className="center">
      <Translate value="debate.thread.noPostsInThread" />
    </div>
  );
};

class Idea extends React.Component {
  constructor(props) {
    super(props);
    this.getTopPosts = this.getTopPosts.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.ideaWithPostsData.idea !== this.props.ideaWithPostsData.idea) {
      this.updateContentLocaleMappingFromProps(nextProps);
    }
  }

  updateContentLocaleMappingFromProps(props) {
    const { defaultContentLocaleMapping, ideaWithPostsData, updateContentLocaleMapping } = props;
    if (!ideaWithPostsData.loading) {
      const postsEdges = ideaWithPostsData.idea.posts.edges;
      const contentLocaleMappingData = {};
      postsEdges.forEach((edge) => {
        const post = edge.node;
        const { id, originalLocale } = post;
        const contentLocale = defaultContentLocaleMapping.get(originalLocale, originalLocale);
        contentLocaleMappingData[id] = {
          contentLocale: contentLocale,
          originalLocale: post.originalLocale
        };
      });

      updateContentLocaleMapping(contentLocaleMappingData);
    }
  }

  getInitialRowIndex = (topPosts, edges) => {
    const { hash } = window.location;
    if (hash !== '') {
      const id = hash.replace('#', '');
      const allPosts = {};
      edges.forEach((e) => {
        allPosts[e.node.id] = e.node;
      });
      let post = allPosts[id];
      if (!post) {
        return null;
      }

      while (post.parentId) {
        post = allPosts[post.parentId];
      }
      const topPostId = post.id;
      const index = topPosts.findIndex((value) => {
        return value.id === topPostId;
      });
      if (index > -1) {
        return index;
      }
      return null;
    }
    return null;
  };

  getTopPosts() {
    const { ideaWithPostsData, routerParams, debateData } = this.props;
    if (!ideaWithPostsData.idea) return [];
    const topPosts = transformPosts(ideaWithPostsData.idea.posts.edges, ideaWithPostsData.idea.messageColumns, {
      refetchIdea: ideaWithPostsData.refetch,
      ideaId: ideaWithPostsData.idea.id,
      routerParams: routerParams,
      debateData: debateData
    });
    return topPosts;
  }

  render() {
    const { contentLocaleMapping, lang, ideaLoading, ideaWithPostsData } = this.props;
    const refetchIdea = ideaWithPostsData.refetch;
    if (ideaLoading) {
      return (
        <div className="idea">
          <Loader />
        </div>
      );
    }
    const { announcement, id, headerImgUrl, synthesisTitle, title } = this.props;
    const isMultiColumn = ideaWithPostsData.loading ? undefined : ideaWithPostsData.idea.messageColumns.length > 0;
    const messageColumns = ideaWithPostsData.loading ? undefined : ideaWithPostsData.idea.messageColumns;
    const childProps = {
      idea: {
        id: id,
        title: title
      },
      ideaWithPostsData: ideaWithPostsData,
      isUserConnected: getConnectedUserId(),
      contentLocaleMapping: contentLocaleMapping,
      refetchIdea: refetchIdea,
      lang: lang,
      noRowsRenderer: noRowsRenderer,
      messageColumns: messageColumns,
      posts: this.getTopPosts(),
      initialRowIndex: ideaWithPostsData.loading
        ? undefined
        : this.getInitialRowIndex(this.getTopPosts(), ideaWithPostsData.idea.posts.edges)
    };
    const view = isMultiColumn ? <ColumnsView {...childProps} /> : <ThreadView {...childProps} />;
    return (
      <div className="idea">
        <Header title={title} synthesisTitle={synthesisTitle} imgUrl={headerImgUrl} identifier="thread" />
        <section className="post-section">
          {!ideaWithPostsData.loading &&
            announcement &&
            <Grid fluid className="background-light">
              <div className="max-container">
                <div className="content-section">
                  <Announcement ideaWithPostsData={ideaWithPostsData} announcementContent={announcement} />
                </div>
              </div>
            </Grid>}
          {ideaWithPostsData.loading ? <Loader /> : view}
        </section>
        <GoUp />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    contentLocaleMapping: state.contentLocale,
    debateData: state.debate.debateData,
    defaultContentLocaleMapping: state.defaultContentLocaleMapping,
    lang: state.i18n.locale
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateContentLocaleMapping: (info) => {
      return dispatch(updateContentLocale(info));
    }
  };
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(IdeaWithPostsQuery, { name: 'ideaWithPostsData' }),
  graphql(IdeaQuery, {
    options: { notifyOnNetworkStatusChange: true },
    // ideaData.loading stays to true when switching interface language (IdeaQuery is using lang variable)
    // This is an issue in apollo-client, adding notifyOnNetworkStatusChange: true is a workaround,
    // downgrading to apollo-client 1.8.1 should works too.
    // See https://github.com/apollographql/apollo-client/issues/1186#issuecomment-327161526
    props: ({ data }) => {
      if (data.loading) {
        return {
          ideaLoading: true
        };
      }
      if (data.error) {
        return {
          ideaHasErrors: true
        };
      }

      return {
        announcement: data.idea.announcement,
        id: data.idea.id,
        title: data.idea.title,
        synthesisTitle: data.idea.synthesisTitle,
        headerImgUrl: data.idea.img ? data.idea.img.externalUrl : ''
      };
    }
  })
)(Idea);