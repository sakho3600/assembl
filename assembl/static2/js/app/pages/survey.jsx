// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { browserHistory } from 'react-router';
import { Translate } from 'react-redux-i18n';
import { Grid, Button } from 'react-bootstrap';
import type { Map } from 'immutable';

import { updateContentLocale } from '../actions/contentLocaleActions';
import withLoadingIndicator from '../components/common/withLoadingIndicator';
import Media from '../components/common/media';
import Header from '../components/debate/common/header';
import Question from '../components/debate/survey/question';
import Navigation from '../components/debate/survey/navigation';
import Proposals from '../components/debate/survey/proposals';
import { getIfPhaseCompletedByIdentifier } from '../utils/timeline';
import ThematicQuery from '../graphql/ThematicQuery.graphql';
import { displayAlert } from '../utils/utilityManager';

type PostNode = {
  node: {
    id: string,
    originalLocale: string
  }
};

type QuestionType = {
  id: string,
  posts: {
    edges: Array<PostNode>
  },
  title: string
};

type SurveyProps = {
  debate: {
    debateData: {
      timeline: string
    }
  },
  defaultContentLocaleMapping: Map,
  hasErrors: boolean,
  imgUrl: string,
  loading: boolean,
  media: Object, // TODO: we should add a type for media/video and use it everywhere
  questions: Array<QuestionType>,
  refetchThematic: Function,
  title: string,
  updateContentLocaleMapping: Function
};

type SurveyState = {
  isScroll: boolean,
  moreProposals: boolean,
  questionIndex: number | null,
  showModal: boolean
};

class Survey extends React.Component<*, SurveyProps, SurveyState> {
  props: SurveyProps;
  state: SurveyState;
  unlisten: Function;

  constructor(props) {
    super(props);
    this.state = {
      isScroll: false,
      moreProposals: false,
      showModal: false,
      questionIndex: null
    };
  }

  componentWillMount() {
    this.updateContentLocaleMappingFromProps(this.props);
  }

  componentDidMount() {
    this.unlisten = browserHistory.listen(() => {
      this.setState({ moreProposals: false });
    });
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.questions !== this.props.questions) {
      this.updateContentLocaleMappingFromProps(nextProps);
    }
  }

  componentWillUnmount() {
    this.unlisten();
  }

  updateContentLocaleMappingFromProps(props) {
    const { defaultContentLocaleMapping, updateContentLocaleMapping, questions } = props;
    const contentLocaleMappingData = {};
    questions.forEach((question) => {
      question.posts.edges.forEach((edge) => {
        const post = edge.node;
        const { id, originalLocale } = post;
        const contentLocale = defaultContentLocaleMapping.get(originalLocale, originalLocale);
        contentLocaleMappingData[id] = {
          contentLocale: contentLocale,
          originalLocale: originalLocale
        };
      });
    });

    updateContentLocaleMapping(contentLocaleMappingData);
  }

  getIfProposals = (questions) => {
    if (!questions) return false;
    let isProposals = false;
    questions.forEach((question) => {
      if (question.posts.edges.length > 0) isProposals = true;
    });
    return isProposals;
  };

  showMoreProposals = () => {
    this.setState({
      moreProposals: true
    });
  };

  scrollToQuestion = (isScroll, questionIndex) => {
    this.setState({
      isScroll: isScroll,
      questionIndex: questionIndex
    });
  };

  render() {
    if (this.props.hasErrors) {
      displayAlert('danger', 'An error occured, please reload the page');
      return null;
    }
    const { imgUrl, media, questions, refetchThematic, title } = this.props;
    const { debateData } = this.props.debate;
    const isPhaseCompleted = getIfPhaseCompletedByIdentifier(debateData.timeline, 'survey');
    return (
      <div className="survey">
        <div className="relative">
          <Header title={title} imgUrl={imgUrl} identifier="survey" />
          <Media {...media} />
          <div className="questions">
            {questions &&
              questions.map((question, index) => {
                return (
                  <Question
                    title={question.title}
                    index={index + 1}
                    key={index}
                    questionId={question.id}
                    scrollToQuestion={this.scrollToQuestion}
                    refetchTheme={refetchThematic}
                  />
                );
              })}
          </div>
          {questions &&
            <Navigation
              questionsLength={questions.length}
              questionIndex={this.state.questionIndex}
              isScroll={this.state.isScroll}
              scrollToQuestion={this.scrollToQuestion}
            />}
          <div className="proposals">
            <section className={isPhaseCompleted ? 'shown' : 'proposals-section'} id="proposals">
              <Grid fluid className="background-light">
                <div className="max-container">
                  <div className="question-title">
                    <div className="title-hyphen">&nbsp;</div>
                    <h1 className="dark-title-1">
                      <Translate value="debate.survey.proposalsTitle" />
                    </h1>
                  </div>
                  <div className="center">
                    {questions &&
                      questions.map((question, index) => {
                        return (
                          <Proposals
                            title={question.title}
                            posts={question.posts.edges}
                            moreProposals={this.state.moreProposals}
                            questionIndex={index + 1}
                            key={index}
                            refetchTheme={refetchThematic}
                          />
                        );
                      })}
                    {!this.state.moreProposals &&
                      this.getIfProposals(questions) &&
                      <Button className="button-submit button-dark" onClick={this.showMoreProposals}>
                        <Translate value="debate.survey.moreProposals" />
                      </Button>}
                  </div>
                  <div className="margin-xl">&nbsp;</div>
                </div>
              </Grid>
            </section>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    debate: state.debate,
    defaultContentLocaleMapping: state.defaultContentLocaleMapping,
    lang: state.i18n.locale
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateContentLocaleMapping: (data) => {
      return dispatch(updateContentLocale(data));
    }
  };
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(ThematicQuery, {
    props: ({ data }) => {
      if (data.loading) {
        return {
          loading: true
        };
      }

      if (data.error) {
        return {
          hasErrors: true
        };
      }

      const { thematic: { img, questions, refetch, title, video: media } } = data;

      return {
        hasErrors: false,
        imgUrl: img.externalUrl,
        loading: false,
        media: media,
        questions: questions,
        refetchThematic: refetch,
        title: title
      };
    }
  }),
  withLoadingIndicator({ color: 'black' })
)(Survey);