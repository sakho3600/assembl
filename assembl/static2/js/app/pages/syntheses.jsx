import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Grid } from 'react-bootstrap';
import { Link } from 'react-router';

import { get } from '../utils/routeMap';
import Loader from '../components/common/loader';
import SynthesesQuery from '../graphql/SynthesesQuery.graphql';

export class DumbSyntheses extends React.Component {
  render() {
    const { data, slug } = this.props;
    if (data.loading) {
      return <Loader color="black" />;
    }
    const { syntheses } = data;
    return (
      <Grid fluid className={syntheses.length > 0 ? 'thumbnails-nav no-padding' : 'hidden'}>
        <div className="thumbnails-container">
          <div className="max-container">
            <div className="thumbnails">
              {syntheses.map((synthesis, index) => {
                return (
                  <div className="thumb-img-container" key={index}>
                    <Link to={`${get('synthesis', { synthesisId: synthesis.id, slug: slug })}`}>
                      <div
                        className={'thumb-img active'}
                        style={synthesis.imgUrl ? { backgroundImage: `url(${synthesis.imgUrl})` } : null}
                      />
                      <div className="color-box">&nbsp;</div>
                      <div className="thumb-title">
                        <div className={'thumb-title-inner'}>
                          {synthesis.subject}
                        </div>
                      </div>
                    </Link>
                  </div>
                );
              })}
            </div>
          </div>
        </div>
      </Grid>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    lang: state.i18n.locale,
    slug: state.debate.debateData.slug
  };
};

export default compose(connect(mapStateToProps), graphql(SynthesesQuery))(DumbSyntheses);