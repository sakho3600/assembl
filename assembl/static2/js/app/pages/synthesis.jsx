// @flow
import React from 'react';
import { Grid, Col } from 'react-bootstrap';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';

import Loader from '../components/common/loader';
import SynthesisQuery from '../graphql/SynthesisQuery.graphql';

export class DumbSynthesis extends React.Component {
  render() {
    const { data } = this.props;
    if (data.loading) {
      return <Loader color="black" />;
    }
    const { synthesis } = data;
    return (
      <Grid fluid>
        <div className="max-container">
          <Col xs={12} sm={12}>
            <div>
              {synthesis.subject}
            </div>
          </Col>
        </div>
      </Grid>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    lang: state.i18n.locale
  };
};

export default compose(
  connect(mapStateToProps),
  graphql(SynthesisQuery, {
    options: (props) => {
      return {
        variables: { id: props.params.synthesisId }
      };
    }
  })
)(DumbSynthesis);