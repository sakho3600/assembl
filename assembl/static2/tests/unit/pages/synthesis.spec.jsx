import React from 'react';
import renderer from 'react-test-renderer';

import { DumbSynthesis } from '../../../js/app/pages/synthesis';

describe('Syntheses component', () => {
  it('should match Syntheses snapshot', () => {
    const props = {
      data: {
        loading: false,
        synthesis: {
          id: 'fooId',
          subject: 'Foo',
          imgUrl: 'http://foo.com/bar'
        }
      }
    };
    const rendered = renderer.create(<DumbSynthesis {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});