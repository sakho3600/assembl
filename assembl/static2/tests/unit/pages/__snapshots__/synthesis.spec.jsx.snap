// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Syntheses component should match Syntheses snapshot 1`] = `
<div
  className="container-fluid"
>
  <div
    className="max-container"
  >
    <div
      className="col-sm-12 col-xs-12"
    >
      <div>
        Foo
      </div>
    </div>
  </div>
</div>
`;
